elastica/rod/factory_function.py

```diff
@@ -1,30 +1,64 @@
 __doc__ = """ Factory function to allocate variables for Cosserat Rod"""
-from typing import Optional, Tuple
+from typing import Any, Optional, Tuple
 import logging
 import numpy as np
 from numpy.testing import assert_allclose
+from numpy.typing import NDArray
 from elastica.utils import MaxDimension, Tolerance
 from elastica._linalg import _batch_cross, _batch_norm, _batch_dot
 
 
 def allocate(
-    n_elements,
-    direction,
-    normal,
-    base_length,
-    base_radius,
-    density,
-    youngs_modulus: float,
+    n_elements: int,
+    direction: NDArray[np.float64],
+    normal: NDArray[np.float64],
+    base_length: np.float64,
+    base_radius: np.float64,
+    density: np.float64,
+    youngs_modulus: np.float64,
     *,
     rod_origin_position: np.ndarray,
     ring_rod_flag: bool,
-    shear_modulus: Optional[float] = None,
+    shear_modulus: Optional[np.float64] = None,
     position: Optional[np.ndarray] = None,
     directors: Optional[np.ndarray] = None,
     rest_sigma: Optional[np.ndarray] = None,
     rest_kappa: Optional[np.ndarray] = None,
-    **kwargs,
-):
+    **kwargs: Any,
+) -> tuple[
+    int,
+    NDArray[np.float64],
+    NDArray[np.float64],
+    NDArray[np.float64],
+    NDArray[np.float64],
+    NDArray[np.float64],
+    NDArray[np.float64],
+    NDArray[np.float64],
+    NDArray[np.float64],
+    NDArray[np.float64],
+    NDArray[np.float64],
+    NDArray[np.float64],
+    NDArray[np.float64],
+    NDArray[np.float64],
+    NDArray[np.float64],
+    NDArray[np.float64],
+    NDArray[np.float64],
+    NDArray[np.float64],
+    NDArray[np.float64],
+    NDArray[np.float64],
+    NDArray[np.float64],
+    NDArray[np.float64],
+    NDArray[np.float64],
+    NDArray[np.float64],
+    NDArray[np.float64],
+    NDArray[np.float64],
+    NDArray[np.float64],
+    NDArray[np.float64],
+    NDArray[np.float64],
+    NDArray[np.float64],
+    NDArray[np.float64],
+    NDArray[np.float64],
```

Comment:
This looks hectic. Can we just use `tuple` or `tuple[Any, ...]`

Reply:
No, we should avoid using `Any` as much as possible. It is really just what it is.

I guess best alternative is removing the first `int` somehow and make it `tuple[NDArray[np.float64], ...]`, but it is also technically not true because it doesn't return variable-sized tuple.